plugins/reindex/src/main/java/org/elasticsearch/plugin/reindex/AbstractAsyncBulkByScrollAction.java

```diff
@@ -0,0 +1,352 @@
+                    IndexResponse ir = item.getResponse();
+                    if (ir.isCreated()) {
+                        created.incrementAndGet();
+                    } else {
+                        updated.incrementAndGet();
+                    }
+                    break;
+                case "delete":
+                    deleted.incrementAndGet();
+                    break;
+                default:
+                    throw new IllegalArgumentException("Unknown op type:  " + item.getOpType());
+                }
+                // Track the indexes we've seen so we can refresh them if requested
+                destinationIndicesThisBatch.add(item.getIndex());
+            }
+            if (destinationIndicesThisBatch.isEmpty() == false) {
+                destinationIndices.addAll(destinationIndicesThisBatch);
+            }
+
+            if (failures.isEmpty() == false) {
+                startNormalTermination();
+                return;
+            }
+
+            if (mainRequest.size() != -1 && successfullyProcessed() >= mainRequest.size()) {
+                // We've processed all the requested docs.
+                startNormalTermination();
+                return;
+            }
+            startNextScrollRequest();
+        } catch (Throwable t) {
+            finishHim(t);
+        }
+    }
+
+    void startNextScrollRequest() {
+        SearchScrollRequest request = new SearchScrollRequest(mainRequest);
+        request.scrollId(scroll.get()).scroll(firstSearchRequest.scroll());
+        client.searchScroll(request, new ActionListener<SearchResponse>() {
+            @Override
+            public void onResponse(SearchResponse response) {
+                onScrollResponse(response);
+            }
+
+            @Override
+            public void onFailure(Throwable e) {
+                finishHim(e);
+            }
+        });
+    }
+
+    private void recordFailure(Failure failure) {
+        switch (failure.getStatus()) {
+        case CONFLICT:
+            versionConflicts.incrementAndGet();
+            if (mainRequest.abortOnVersionConflict()) {
+                failures.add(failure);
+            }
+            return;
+        default:
+            failures.add(failure);
+        }
+    }
+
+    void startNormalTermination() {
+        if (mainRequest.refresh()) {
+            RefreshRequest refresh = new RefreshRequest(mainRequest);
+            refresh.indices(destinationIndices.toArray(new String[destinationIndices.size()]));
+            client.admin().indices().refresh(refresh, new ActionListener<RefreshResponse>() {
+                @Override
+                public void onResponse(RefreshResponse response) {
+                    finishHim(null);
+                }
+
+                @Override
+                public void onFailure(Throwable e) {
+                    finishHim(e);
+                }
+            });
+            return;
+        }
+        finishHim(null);
```

Comment:
I think this should be in an else branch, as we're calling this method twice when `mainRequest.refresh() == true` (once from the callback and once here).